staging/src/k8s.io/sample-controller/artifacts/examples/main.go

```diff
@@ -0,0 +1,129 @@
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/scale"
+	"k8s.io/client-go/tools/clientcmd"
+	clientset "k8s.io/sample-controller/pkg/client/clientset/versioned"
+)
+
+const (
+	namespace       = "default"
+	resourceName    = "example-foo"
+	resourceGroup   = "samplecontroller.k8s.io"
+	resourceVersion = "v1alpha1"
+)
+
+func main() {
+	err := start()
+	if err != nil {
+		glog.Fatalf("error: %v", err)
+		os.Exit(1)
+	}
+
+}
+func start() error {
+	var cfg *string
+	if c := kubeconfig(); c != "" {
+		cfg = flag.String("kubeconfig", c, "(optional) KUBECONFIG env of the kubeconfig file")
+	} else {
+		cfg = flag.String("kubeconfig", "", "absolute path to the kubeconfig file")
+	}
+	flag.Parse()
+
+	config, err := clientcmd.BuildConfigFromFlags("", *cfg)
+	exampleClient, err := clientset.NewForConfig(config)
+	if err != nil {
+		glog.Fatalf("error build example clientset: %s", err.Error())
+		return err
+	}
+	_, err = exampleClient.SamplecontrollerV1alpha1().Foos(namespace).Get(resourceName, metav1.GetOptions{})
+	if err != nil {
+		glog.Fatalf("err get example-foo: %s", err.Error())
+		return err
+	}
+	fooScaleClient, err := createNewScaleClient(config)
+	if err != nil {
+		glog.Fatalf("error build scale client: %s", err.Error())
+	}
+	groupResource := schema.GroupResource{
+		Group:    resourceGroup,
+		Resource: "foos",
+	}
+	gottenScale, err := fooScaleClient.Scales(namespace).Get(groupResource, resourceName)
+	if err != nil {
+		glog.Fatalf("error get the scale")
+		return nil
+	}
+	glog.Infof("Got the scale subresource: %v", gottenScale)
+	currentNo := gottenScale.Spec.Replicas
+	gottenScale.Spec.Replicas = currentNo + 1
+	fooScaleClient.Scales(namespace).Update(groupResource, gottenScale)
+	glog.Infof("Scale the replica from %v to %v using scale subresource", currentNo, gottenScale.Spec.Replicas)
+	return nil
+}
+
+func kubeconfig() string {
+	if h := os.Getenv("KUBECONFIG"); len(h) > 0 {
+		return h
+	}
+	return ""
+}
+
+func createNewScaleClient(config *rest.Config) (scale.ScalesGetter, error) {
+	discoveryClient, err := discovery.NewDiscoveryClientForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+	groupResource, err := discoveryClient.ServerResourcesForGroupVersion(resourceGroup + "/" + resourceVersion)
+	if err != nil {
+		return nil, err
+	}
+
+	resources := []*discovery.APIGroupResources{
+		{
+			Group: metav1.APIGroup{
+				Name: resourceGroup,
+				Versions: []metav1.GroupVersionForDiscovery{
+					{Version: resourceVersion},
+				},
+				PreferredVersion: metav1.GroupVersionForDiscovery{Version: resourceVersion},
+			},
+			VersionedResources: map[string][]metav1.APIResource{
+				resourceVersion: groupResource.APIResources,
+			},
+		},
+	}
+
+	restMapper := discovery.NewRESTMapper(resources, nil)
+	resolver := scale.NewDiscoveryScaleKindResolver(discoveryClient)
```

Comment:
Can we separate obtaining the discovery info into its own function, and add a comment explaining that is what we are doing and why.